Make .to_dict() output the default, misc. other API improvements

valgur opened the pull request:
Comparing the `.to_dict()` output to the default `.query()` output, I have a hard time imagining anyone being interested in using the raw JSON response from the server. Because of this, I propose making the dictionary output the default for `.query()`, which this PR implements.

`.query()` now returns a parsed dictionary, while the `.load_query()` still returns the raw JSON and has been made a private method (prefixed with an underscore). To support full-text searches in a convenient manner, I added `.query_plain()` that takes just a query string and returns a parsed dictionary as well.

I also modified the dictionary structure slightly. I made the product ID is now the dictionary key instead of the product title. This is because of non-uniqueness of the product ID as the identifier. There are quite a few duplicate products with different IDs in the Data Hub (e.g. bit-by-bit identical products ending in '_0135' in [this query](https://scihub.copernicus.eu/apihub/search?q=%28beginPosition%3A%5B2014-01-01T00%3A00%3A00Z+TO+2014-12-31T00%3A00%3A00Z%5D%29+AND+%28footprint%3A%22Intersects%28POLYGON%28%28-66.2695312+-8.0592296%2C-66.2695312+0.7031074%2C-57.3046875+0.7031074%2C-57.3046875+-8.0592296%2C-66.2695312+-8.0592296%29%29%29%22%29+AND+%28producttype%3AGRD%29&rows=100&format=json)) and only one would be retained in the query output. Dropping duplicates would actually be very useful functionality as it would prevent unnecessary downloads, but I think this is not a good spot for doing that. We could consider adding an additional function for taking care of duplicates.

In addition to that major change I also updated the download code to expect a list of product IDs as input (so the parsed dictionaries work directly as well). This makes it easy to pass just a `DataFrame` index consisting of product IDs as the input to `.download_all()` as well. I also fixed the download error reporting bugs on the CLI interface reported in #88 and added unit tests to cover the CLI downloading code. This brings the unit test code coverage to near 100%. Only rare exception handling remains uncovered, which is perfectly fine.

I also added the improvement ideas in #101, although I seem to have missed @Fernerkundung's last comment.

There is still some work left to do regarding this PR. I have not updated the documentation yet, for one. I wanted to solicit your opinions first before doing that. Do you agree with the proposed changes?
I'm not sure about the method name for `query_plain`, for example. I think it could be named batter to reflect its purpose of doing (plain) full-text search.

valgur commented:
The first few commits are from #105, which has not been merged yet.

Fernerkundung commented on tests/test_mod.py:
```diff
@@ -287,25 +325,30 @@ def test_footprints_s1():
+def test_footprints_s2(products):
+    footprints = SentinelAPI.to_geojson(products)
+    for footprint in footprints['features']:
+        print(footprint)
```
I don't think printing messages NOT from pytest is a good idea in this context - I think this might catch future contributors by surprise.

valgur replied:
Thanks for catching this. This was meant to be temporary, to help with debugging.

Fernerkundung commented:
Great contribution @valgur!
`query_plain` is just a more convenient shortcut to `_response_to_dict(_load_query(<query>))` but I think a good idea. I'd favour your intial naming convention of this as `query_raw` as the query is not expected in plain(text) but rather as a formated opensearch query.

Regarding the duplicate products with differing `UUID` - I don't think we should stress testing for this inside `sentinelsat`. This is a bug inside the SciHub server/database integration and thus should be dealt with on the server side. If we want to include this information I'd say it is best located in the Wiki/Docs, rather than as a further error handling mechanism.

The split of `get_coordinates` into the read and the formatting part I suggested in #101 can also be dealt with in a later PR. I still see it as a valid addition, because with this we could make the API and CLI 'geoJSON native' in that users can directly (re)use their JSON features they created with other tools such as rasterio, fiona or even other APIs.

👍 👍  for enabling `download_all(filtered_dataframe.index)`! A lot more straight forward than the current implementation.

Fernerkundung commented:
@valgur Do you think this is ready to merge? If so I will build #54 on top of it to make merging easier for @willemarcel

valgur commented:
No, I would not. The latest commits are quite iffy. I had some ideas in mind about refactoring the downloading code, somewhat in preparation for #91, but I don't think it's a good approach anymore. I think I'll revert most of the later changes for now except for the extended OData querying functionality, which I think is very useful. Documentation needs updating, too. I'll see if I can fix that today.

valgur commented:
I modified the last commits and kept only the changes I'm happy with. I also documented my changes in the changelog and updated the docs in general.
That added [changelog entry](https://github.com/valgur/sentinelsat/blob/api-improvements/CHANGELOG.rst#010--2017-xx-xx) is also a good summary of the changes introduced by this PR.

The pull request should be ready for merging now. Feel free to let me know if you don't quite agree with anything, though.

valgur commented:
I find the (now quite powerful) metadata querying capability of `sentinelsat` to be a significant and useful feature which differentiates us from other similar utilities, so I suggest updating our tagline in the README to reflect that:

> Sentinelsat makes searching, retrieving the metadata and downloading of [Sentinel](http://www.esa.int/Our_Activities/Observing_the_Earth/Copernicus/Overview4)
> satellite images from the [Copernicus Open Access Hub](https://scihub.copernicus.eu/) easy.

Do you agree?

Fernerkundung commented:
+1 on changing the tagline. The ability to fine tune queries and now also the local filtering/analysis of the metadata is something I haven't seen in another library.

Great changes overall @valgur!

j08lue commented:
You are putting a lot of very good work into this, @valgur.
